README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

Generated RSA pub/priv key with a seed string (Using bip39 isn't a bad idea ;-)

## Features

To secure things, RSA key pair is still a good idea, but it's not really user friendly
Handling it with the same idea as bitcoin do, generating it from a uuid size seed.

Key pair is generated by 2048 deterministic derivation using sha512
Then creating every BigInt needed by writing in a string the next derivation until the string fit our key size need

## Usage

Creating the generator with a seed:
~~~
final seededRSA = SeededRSA(mnemonnic);
~~~

Calling the generating process
~~~
await seededRSA.generate()
~~~

## Disclaimer

This library isn't tested or review
Do not used it, or do the review/evaluation yourself ;-)

Starded by porting https://github.com/nomadcrypto/seededrsa but finally made totaly diffrently ;-)


CHANGELOG.md
## 1.0.0

- First almost Working Version

## 1.0.1

- Correction on failed pubkey generation

## 1.0.2

-deleting RSA from PEM head and foot on public key, some implementation care about details !


test/bip39_seeded_rsa_test.dart
import 'package:bip39_seeded_rsa/bip39_seeded_rsa.dart';
//import 'package:test/test.dart';
import 'package:bip39/bip39.dart' as bip39;
import 'package:ninja_prime/ninja_prime.dart';

import 'package:crypto/crypto.dart';
import 'dart:convert';
import 'dart:math';

void main() async {
  String mnemonnic = "wash list guard stool slogan lift boss imitate story trash put option";
  //bip39.generateMnemonic();
  final awesome = RSAKey(mnemonnic);

  //print(SeededRandom(mnemonnic).randomPrimeBigInt(2048));
  print(await awesome.generate());


}


lib/src/seeded_random.dart
import 'package:crypto/crypto.dart';
import 'package:bip39/bip39.dart' as bip39;
import 'dart:math';
import 'package:ninja_prime/ninja_prime.dart';

class SeededRandom {
  late Digest digest;
  SeededRandom(String mnemonic) {
    if(bip39.validateMnemonic(mnemonic)){
      digest = sha512.convert(bip39.mnemonicToSeed(mnemonic));
    } else {
      digest = sha512.convert(mnemonic.codeUnits);
    }

    //based derivations
    for(var i = 0 ; i < 2048; i++) {
      digest = sha512.convert(digest.bytes);
    }
  }

  double logBase(num x, num base) => log(x) / log(base);
  double log2(num x) => logBase(x, 2);

  int getKeySizeInDecimal(int keysize) => -(-(keysize * log(2)/log(10))).floor();

  BigInt _getNextBigInt(int size) {
    var bigIntStr = digest.bytes.join();
    while(bigIntStr.length < size){
      digest = sha512.convert(digest.bytes);
      bigIntStr += digest.bytes.join();
    }

    bigIntStr = bigIntStr.substring(0,size);
    return BigInt.parse(bigIntStr);
  }

  BigInt randomPrimeBigInt(int keysize){
    int size = getKeySizeInDecimal(keysize);
    var bi = _getNextBigInt(size);

    while(!bi.isPrime()){
      bi = _getNextBigInt(size);
    }
    return bi;
  }
}


lib/src/seeded_rsa_base.dart
import 'dart:convert';
import 'package:ninja_asn1/ninja_asn1.dart';
import 'package:seeded_rsa/src/seeded_random.dart';

class SeededRSA {
  static const int PEM_CHAR_LINE_SIZE = 64;
  late BigInt n, d, p, q, dmp1, dmq1, coeff, e = BigInt.zero;
  late String seed;

  SeededRSA(this.seed);

  String _toPem(String pkeyb64, {String key = "RSA PRIVATE KEY"}){
    String pemString = "-----BEGIN $key-----\n";
    for(var i = 0; i < pkeyb64.length; i+=PEM_CHAR_LINE_SIZE) {
      var i64 = i+PEM_CHAR_LINE_SIZE;
      if(i64 > pkeyb64.length) i64 = pkeyb64.length;
      pemString+="${pkeyb64.substring(i, i64)}\n";
    }
    pemString += "-----END $key-----";
    return pemString;
  }

   String privateKey() {
    final original = ASN1Sequence(
      [
      ASN1Integer(BigInt.zero),
      ASN1Integer(n),
      ASN1Integer(e),
      ASN1Integer(d),
      ASN1Integer(p),
      ASN1Integer(q),
      ASN1Integer(dmp1),
      ASN1Integer(dmq1),
      ASN1Integer(coeff)
      ]
    );
    return _toPem(base64Encode(original.encode()));
  }

  String publicBaseKey(){
    var first_sequence = ASN1Sequence([
        ASN1ObjectIdentifier.fromString('1.2.840.113549.1.1.1'),
        ASN1Null(),
        ]);

    var secondSequence = ASN1Sequence([
        ASN1Integer(n),
        ASN1Integer(e)
      ]);

    var bitString = ASN1BitString(secondSequence.encode());

    var seq = ASN1Sequence([
        first_sequence,
        bitString
      ]);

    return _toPem(base64Encode(seq.encode()), key: "PUBLIC KEY");
  }


  Future<Map<String,String>> generate({int keySize = 2048, String exposant = "65537"}) async {
    var seededRandom = SeededRandom(seed);
    var qs = keySize >> 1;
    e = BigInt.parse(exposant,radix: 16);
    var exponent = BigInt.parse(exposant, radix: 16);

    var valid = false;

    while(!valid) {
        try{
          p = seededRandom.randomPrimeBigInt(keySize-qs);
          q = seededRandom.randomPrimeBigInt(qs);
          if(p < q){
            var t = p;
            p = q;
            q = p;
          }

          var p1 = p - BigInt.one;
          var q1 = q - BigInt.one;
          var phi = p1 * q1;
          if(phi.gcd(exponent) == BigInt.one){
            n = p * q;
            d = exponent.modInverse(phi);
            dmp1 = d % p1;
            dmq1 = d % q1;
            coeff = q.modInverse(p);
            valid = true;
          }
        } catch(e){
          valid = false;
        }
    }

    var key = {
      'privateKey': privateKey(),
      'publicKey': publicBaseKey()
    };

    return key;
  }


}


lib/src/bip39_seeded_rsa_base.dart
import 'dart:math';
import 'dart:convert';
import 'dart:typed_data';
import 'package:crypto/crypto.dart';
import 'package:ninja_prime/ninja_prime.dart';
import 'package:ninja_asn1/ninja_asn1.dart';
import 'package:bip39_seeded_rsa/src/seeded_random.dart';

class RSAKey {
  static const int PEM_CHAR_LINE_SIZE = 64;
  late BigInt n, d, p, q, dmp1, dmq1, coeff, e = BigInt.zero;
  late String seed;
  late SeededRandom seededRandom = SeededRandom(seed);

  RSAKey(this.seed);


  BigInt _getPrime(int keySize){
    var r = seededRandom.randomPrimeBigInt(keySize);

    while(!r.isPrime()){
      r = seededRandom.randomPrimeBigInt(keySize);
    }
    return r;
  }

  String _toPem(String pkeyb64, {String key = "RSA PRIVATE KEY"}){
    String pemString = "-----BEGIN $key-----\n";
    for(var i = 0; i < pkeyb64.length; i+=PEM_CHAR_LINE_SIZE) {
      var i64 = i+PEM_CHAR_LINE_SIZE;
      if(i64 > pkeyb64.length) i64 = pkeyb64.length;
      pemString+=pkeyb64.substring(i, i64) + "\n";
    }
    pemString += "-----END $key-----";
    return pemString;
  }

   String privateKey() {
    final original = ASN1Sequence(
      [
      ASN1Integer(BigInt.zero),
      ASN1Integer(n),
      ASN1Integer(e),
      ASN1Integer(d),
      ASN1Integer(p),
      ASN1Integer(q),
      ASN1Integer(dmp1),
      ASN1Integer(dmq1),
      ASN1Integer(coeff)
      ]
    );
    return _toPem(base64Encode(original.encode()));
  }

  String publicBaseKey() {
    var second_sequence = ASN1Sequence(
      [
        ASN1Integer(n),
        ASN1Integer(e)
      ]
      );

    String sq =  "00" + base64Encode(second_sequence.encode()).toString();
    var seq = ASN1Sequence(
      [
        ASN1ObjectIdentifier.fromString('1.2.840.113549.1.1.1'),
        ASN1Null(),
        ASN1BitString(Uint8List.fromList(sq.codeUnits))
      ]
      );

    return _toPem(base64Encode(seq.encode()), key: "RSA PUBLIC KEY");
  }


  Future<Map<String,String>> generate({int keySize = 2048, String exposant = "65537"}) async {
    var qs = keySize >> 1;
    e = BigInt.parse(exposant,radix: 16);
    var exponent = BigInt.parse(exposant, radix: 16);

    var valid = false;

    while(!valid) {
      try {
        p = _getPrime(keySize);
        q = _getPrime(keySize);
        if(p < q){
          var t = p;
          p = q;
          q = p;
        }

        var p1 = p - BigInt.one;
        var q1 = q - BigInt.one;
        var phi = p1 * q1;
        if(phi.gcd(exponent) == BigInt.one){
          n = p * q;
          d = exponent.modInverse(phi);
          dmp1 = d % p1;
          dmq1 = d % q1;
          coeff = q.modInverse(p);
          valid = true;
        }
      } catch (e){
        valid = false;
      }
    }

    var key = {
      'privateKey': privateKey(),
      'publicKey': publicBaseKey()
    };

    return key;
  }


}
